### PR TITLE
Refactor the Makefile and hack/ scripts

### PR DESCRIPTION
**Pull request by ConnorJC3**
**Is this a bug fix or adding new feature?**

"New Feature"/Refactoring

**What is this PR about? / Why do we need it?**

This PR contains a significant refactor of the Makefile and e2e testing scripts. Its goals are:

- **Primary Goal: Modularize `run.sh` to enable separate cluster creation, test running, and cluster deletion.** Today, running E2E tests locally requires waiting on the creation and termination of a cluster before and after the run, which takes significant time. Developers should be able to run E2E tests locally against a pre-existing cluster, as well as easily create and delete clusters for testing that follow the standard E2E setup.
- **Primary Goal: Document the available `Makefile` targets and how to use them.** New contributors should not need to go digging in the `Makefile` to discover how to perform simple tasks such as building the driver or running e2e tests.
- **Secondary Goal: Centralize dependency management.** We should have a single place that all local tools are installed - rather than the existing code which has multiple different methods of achieving the same task of installing a binary from GitHub.
- **Secondary Goal: Simply Makefile** - many of the existing makefile targets (in particular, the e2e targets) are long, unwieldy, and hard to reason about. Clean up the makefile and `hack/` directory as much as possible to remove unneeded complexity.
- **Explicit Non-Goal: Significant changes to how the image is built or how the e2e tests are ran.** I do think we should look into improvements in both of these places in the future, but this PR already contains enough changes as is. This PR should not change the behavior of our CI (other than making it run faster and easier to develop for). To this end, `eksctl.sh`, `kops.sh`, and `metrics.sh` are largely or entirely unmodified.
- **Explicit Non-Goal: Changes to Go files** - this PR is only about refactoring the testing infrastructure, not the tests themselves.

**What testing is done?** 

Significant local testing as well as CI, see the new `docs/makefile.md` for the available makefile targets to test against locally

### CODE REVIEW

**Comment by AndrewSirenko** [comment]
Still working through this PR, but I found this visualization helpful

Pre-PR: 
pre-pr-Makefile

Post-PR: 
post-pr-Makefile

Generated with [vak/makefile2dot: Visualize your Makefile using GraphViz dot utility](https://github.com/vak/makefile2dot)

**Comment by AndrewSirenko** [comment]
Can we rebase this PR to separate linting, code extraction, filename changes, and variable name changes into separate commits from the main changes to business logic? 

Perhaps commits in this order:

1. Variable name changes
2. Code extraction 
3. Filename changes
4. *Main Business logic commit* (Bonus points if there are separate commits for Each primary/secondary goal)
5. Linting

Perhaps by running `git reset master` followed by re-committing changes by hunk/File? 

Let me know if parallel ~programming~ committing would help in this endeavor. I did something similar for the original versions of #1846 and #1845 before squashing those commits. 

Also: 
> Significant local testing as well as CI, see the new docs/makefile.md for the available makefile targets to test against locally

Means that you tested each makefile target locally and on CI, correct?

**Comment by ConnorJC3** [comment]
> Can we rebase this PR to separate linting, code extraction, filename changes, and variable name changes into separate commits from the main changes to business logic?

Is there a significant advantage to doing the large amount work to split this into separate commits? They will ultimately need to be squashed for merge (as all of the commits you listed will be in a broken, partial state) and doing so greatly increases the effort required to update the PR for review comments.

> Means that you tested each makefile target locally and on CI, correct?

All the targets were tested locally, and CI runs the targets it already does (`make verify`, `make test`, and all the e2e tests)

File: docs/makefile.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
+# Use of the EBS CSI Driver `Makefile`
+
+The EBS CSI Driver comes with a Makefile that can be used to develop, build, test, and release the driver. This file documents the available targets intended for use by humans, the parameters they support, and common usage scenarios.
```
Comment (torredil) [inline docs/makefile.md:3]: np suggest the following: _The EBS CSI Driver comes with a Makefile that can be used to develop, build, test, and release the driver. This file documents the available targets ~intended for use by humans~, the parameters they support, and common usage scenarios._

**Comment by torredil** [comment]
Hitting a quota limit here:
```
2023-12-11 21:30:31 [✖]  creating OIDC provider: operation error IAM: CreateOpenIDConnectProvider, https response error StatusCode: 409, RequestID: ce25e2a1-fa71-48e4-93a9-0e26ffd23971, LimitExceeded: Cannot exceed quota for OpenIdConnectProvidersPerAccount: 100
```

cc: @dims

**Comment by dims** [comment]
@torredil cleaned up a few, please try again?

File: hack/e2e/config.sh [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+DEPLOY_METHOD=${DEPLOY_METHOD:-"helm"}
+HELM_CT_TEST=${HELM_CT_TEST:-"false"}
+HELM_EXTRA_FLAGS=${HELM_EXTRA_FLAGS:-}
+COLLECT_METRICS=${COLLECT_METRICS:-"false"}
+
+TEST_PATH=${TEST_PATH:-"./tests/e2e-kubernetes/..."}
+GINKGO_FOCUS=${GINKGO_FOCUS:-"External.Storage"}
+GINKGO_SKIP=${GINKGO_SKIP:-"\[Disruptive\]|\[Serial\]"}
+GINKGO_PARALLEL=${GINKGO_PARALLEL:-25}
```
Comment (AndrewSirenko) [inline hack/e2e/config.sh:59]: Referencing `GINKGO` here may cause confusion to whether it is the call of ginkgo by the AWS E2E tests, versus the K8S storage conformance tests. I would recommend renaming the variable in line with the other config constants I suggested in the comments above.

(Though this is a direct upgrade already to what currently exists)

File: hack/e2e/kops/kops.sh [hunk 3]
```diff
@@ -119,15 +116,15 @@ function kops_patch_cluster_file() {
   if [ -n "$ROLE" ]; then
     FILTER="$FILTER | select(.spec.role==\"$ROLE\")"
   fi
-  jq "$FILTER" "$CLUSTER_FILE_JSON" > "$CLUSTER_FILE_0"
+  jq "$FILTER" "$CLUSTER_FILE_JSON" >"$CLUSTER_FILE_0"
 
   # Patch only the json objects
-  kubectl patch -f "$CLUSTER_FILE_0" --local --type merge --patch "$(cat "$KOPS_PATCH_FILE")" -o json > "$CLUSTER_FILE_1"
```
Comment (AndrewSirenko) [inline hack/e2e/kops/kops.sh:125]: Nit: Can you put whatever linter you used for our scripts into the makefile documentation? In a future PR we can add an update target that runs the bash linter.

**Review by AndrewSirenko — CHANGES_REQUESTED** [review]
Great job on this!! *Super* beneficial to onboarding contributors to the project, as well as maintaining maintainer sanity. Ran a few of the commands, will make sure to run through all of them when I find more time. Will start using this makefile for daily driver development and see what else I find.

File: Makefile [hunk 4]
```diff
@@ -12,29 +12,35 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+###
+### How to use this Makefile is documented in docs/makefile.md
+###
+
+## Variables/Functions
+
 VERSION?=v1.25.0
```
[the remark this answers is not in the document]
Comment (ConnorJC3) [reply Makefile:21]: This will be fixed when the PR is rebased

File: Makefile [hunk 5]
```diff
@@ -43,244 +49,110 @@ ALL_OS_ARCH_OSVERSION_linux=$(foreach arch, $(ALL_ARCH_linux), $(foreach osversi
...
-	AWS_REGION=us-west-2 \
+.PHONY: e2e/single-az
+e2e/single-az: bin/helm bin/ginkgo
 	AWS_AVAILABILITY_ZONES=us-west-2a \
-	HELM_EXTRA_FLAGS='--set=controller.k8sTagClusterId=$$CLUSTER_NAME,controller.volumeModificationFeature.enabled=true' \
-	EBS_INSTALL_SNAPSHOT="true" \
 	TEST_PATH=./tests/e2e/... \
 	GINKGO_FOCUS="\[ebs-csi-e2e\] \[single-az\]" \
-	GINKGO_SKIP="\"sc1\"|\"st1\"" \
+	GINKGO_PARALLEL=5 \
```
[the remark this answers is not in the document]
Comment (ConnorJC3) [reply Makefile:115]: These are all non-default values, not sure what you mean?

File: docs/makefile.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+- `git`: https://git-scm.com/downloads
+- `docker` and `docker buildx`: https://docs.docker.com/get-docker/ and https://github.com/docker/buildx#installing
+- `make`
+- Standard POSIX tools (`awk`, `grep`, `cat`, etc)
+
+All other tools are downloaded for you at runtime.
+
+## Quickstart Guide
+
+This guide demonstrates the basic workflow for developing for the EBS CSI Driver. More detailed documentation of the available `make` targets is available below.
```
[the remark this answers is not in the document]
Comment (ConnorJC3) [reply docs/makefile.md:21]: Can you clarify the ask? Do you want me to remove the quick start guide entirely?

File: docs/makefile.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+### 1. Local development for the EBS CSI Driver
+
+If your changes are Helm-only, skip to section 2 (Run E2E tests) below after making your changes.
+
+During development, use `make` at any time to build a driver binary, and thus discover compiler errors. When your change is ready to be tested, run `make test` to execute the unit test suite for the driver. If you are making a significant change to the driver (such as a bugfix or new feature), please add new unit tests or update existing tests where applicable.
+
+### 2. Run E2E tests
+
+To create a `kops` cluster to test the driver against:
```
[the remark this answers is not in the document]
Comment (ConnorJC3) [reply docs/makefile.md:31]: The makefile (and therefore the makefile documentation) is used by contributors, not users, and thus is biased to their needs.

File: hack/e2e/config.sh (hunk 2 above)
Comment (ConnorJC3) [reply hack/e2e/config.sh:59]: We do call gingko directly in some cases, what would be your suggested names?

These are the standard names the old makefile already used.

File: hack/e2e/kops/kops.sh (hunk 3 above)
Comment (ConnorJC3) [reply hack/e2e/kops/kops.sh:125]: It's already in this PR, see target `update/shfmt`

File: Makefile [hunk 8]
```diff
@@ -291,4 +163,101 @@ generate-sidecar-tags: update-truth-sidecars charts/aws-ebs-csi-driver/values.ya
...
+
+.PHONY: verify/govet
+verify/govet:
+	go vet $$(go list ./...)
+
+.PHONY: verify/update
+verify/update: bin/helm bin/mockgen
+	./hack/verify-update.sh
+
+## Internal
```
[the remark this answers is not in the document]
Comment (ConnorJC3) [reply Makefile:259]: There used to be more targets than the directory one in a previous version of the PR. Now that there's only one I'll just move that up to the build section

File: docs/makefile.md (hunk 1 above)
Comment (ConnorJC3) [reply docs/makefile.md:3]: Fixed in next version

File: docs/makefile.md [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+The `Makefile` and associated tooling assume you already have the following tools installed:
+- `go`: https://go.dev/doc/install
+- `python` (may be named `python3`) and `pip`: https://www.python.org/downloads/
+- `jq`: https://github.com/jqlang/jq/releases
+- `kubectl`: https://kubernetes.io/docs/tasks/tools/#kubectl
+- `git`: https://git-scm.com/downloads
+- `docker` and `docker buildx`: https://docs.docker.com/get-docker/ and https://github.com/docker/buildx#installing
+- `make`
+- Standard POSIX tools (`awk`, `grep`, `cat`, etc)
```
[the remark this answers is not in the document]
Comment (ConnorJC3) [reply docs/makefile.md:15]: Fixed in next version other than `This` -> `The`

File: Makefile [hunk 10]
```diff
@@ -43,244 +49,110 @@ ALL_OS_ARCH_OSVERSION_linux=$(foreach arch, $(ALL_ARCH_linux), $(foreach osversi
...
 	WINDOWS=true \
-	HELM_VALUES_FILE="./hack/values_eksctl.yaml" \
-	HELM_EXTRA_FLAGS='--set=controller.k8sTagClusterId=$$CLUSTER_NAME' \
-	EKSCTL_ADMIN_ROLE="Infra-prod-KopsDeleteAllLambdaServiceRoleF1578477-1ELDFIB4KCMXV" \
-	AWS_REGION=us-west-2 \
-	AWS_AVAILABILITY_ZONES=us-west-2a,us-west-2b \
-	TEST_PATH=./tests/e2e-kubernetes/... \
-	GINKGO_FOCUS="External.Storage" \
 	GINKGO_SKIP="\[Disruptive\]|\[Serial\]|\[LinuxOnly\]|\[Feature:VolumeSnapshotDataSource\]|\(xfs\)|\(ext4\)|\(block volmode\)" \
 	GINKGO_PARALLEL=15 \
```
[the remark this answers is not in the document]
Comment (ConnorJC3) [reply Makefile:140]: That would require putting magic if statements in run.sh

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (AndrewSirenko) [reply Makefile:113]: Ignore all of these types of comments. My brain failed to realize these are being 'passed into' the command. 😞

**Review by AndrewSirenko — APPROVED** [review]
LGTM once rebased.

**Review by torredil — APPROVED** [review]
/lgtm
/approve
/hold

**Pull request merged**